Add aws terraform configuration

csadorf opened the pull request:
Adds configuration for setting up an AiiDAlab-compatible cluster on AWS resources using terraform.

To be merged after #2.

csadorf committed:
Configure aws deployment with terraform.

csadorf committed:
Extend gitignore.

csadorf commented on aws/terraform/LICENSE:
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) 2020 Pangeo Data
```
@yakutovicha Since I made changes, we might want/have to add a line here?
```suggestion
Copyright (c) 2020 Simon Adorf (EPFL)
Copyright (c) 2020 Pangeo Data
```

csadorf committed:
Update aws/terraform/LICENSE

csadorf commented:
@yakutovicha Is this ready to merge?

mbercx commented on aws/terraform/README.md:
```diff
@@ -0,0 +1,286 @@
+
+## Introduction
+
+This repo houses an opinionated deployment of PANGEO-style
```
General comment: I rather like the convention we have for the AiiDA docs/tutorials of one sentence per line for easy diffs. Is there a reason why you're not doing this here?

csadorf replied:
I did not adapt the README in any substantial way. This is now fixed.

mbercx commented on aws/terraform/README.md:
```diff
@@ -0,0 +1,286 @@
+should help.
+
+This repo provides the `aws-creds` folder in case you do not have
+admin permissions or want to follow the principle of least privilege.
```
You do need to have quite a few permissions on AWS to be able to follow the steps for setting up the `terraform-bot` and assigning it with the policy. Of course, if we decide to go with the one-bot-for-all approach, the one following these instructions might not even need access to AWS in case they have the `terraform-bot` credentials?

csadorf replied:
I would think so, yes.

mbercx commented on aws/terraform/README.md:
```diff
@@ -0,0 +1,286 @@
+- [kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/)
+- [Helm](https://helm.sh/docs/intro/install/)
+
+You will also need this repo. You can get it with:
```
Would it perhaps be a good idea to run terraform remotely on Terraform Cloud or elsewhere for our tutorials?

https://learn.hashicorp.com/tutorials/terraform/aws-remote?in=terraform/aws-get-started

csadorf replied:
I'd prefer to not need to rely on an external provider, but manage the configuration in this repository.

mbercx commented on aws/terraform/README.md:
```diff
@@ -0,0 +1,286 @@
+
+The terraform deployment needs several variable names set before it
+can start. If you look in `aws/your-cluster.tfvars`, there are four
+variables present. You should input cluster and vpc names. You only
```
It's not entirely clear at this point what the VPC and cluster names should be. Can I choose them freely? Do I have to look them up on the AWS console? What the heck is a VPC anyways?

EDIT: In the files of _this_ repo, the file is called `aiidalab-cluster.tfvars`, and the `vpc_name` and `cluster_name` are already filled in.

csadorf replied:
I hope this is now better described in the README.

mbercx commented on aws/terraform/aws-creds/iam.tf:
```diff
@@ -0,0 +1,196 @@
+}
+
+provider "aws" {
+  version = ">= 2.28.1"
```
Seems this usage is deprecated:

```console
Warning: Version constraints inside provider configuration blocks are deprecated

  on main.tf line 6, in provider "aws":
   6:   version = ">= 2.28.1"

Terraform 0.13 and earlier allowed provider version constraints inside the
provider configuration block, but that is now deprecated and will be removed
in a future version of Terraform. To silence this warning, move the provider
version constraint into the required_providers block.

(and 2 more similar warnings elsewhere)
```

You now have to provide the version constraints in the `terraform` block, as detailed [here](https://www.terraform.io/docs/language/providers/requirements.html).

csadorf replied:
This was fixed.

mbercx commented on aws/terraform/aws/main.tf:
```diff
@@ -0,0 +1,204 @@
+terraform {
+  required_version = ">= 0.12.6"
```
EDIT: This whole thread of comments explains me trying to figure out what is going wrong with the versions, while using the blog-post repo. Since some of the issues are also present for this repo, I'll leave this here, but the `helm` provider related issues are not present when actually using the terraform files in this repo.

---

When trying to run `terraform init`, I got the following error:

```console
Error: Unsupported Terraform Core version

  on .terraform/modules/iam_assumable_role_admin/modules/iam-assumable-role-with-oidc/versions.tf line 2, in terraform:
   2:   required_version = "~> 0.12.6"

Module module.iam_assumable_role_admin (from
terraform-aws-modules/iam/aws//modules/iam-assumable-role-with-oidc) does not
support Terraform version 0.14.5. To proceed, either choose another supported
Terraform version or update this version constraint. Version constraints are
normally set for good reason, so updating the constraint may lead to other
errors or unexpected behavior.
```

mbercx replied:
I fixed this by using [`tfenv`](https://stackoverflow.com/a/56283464) to change my terraform version to 0.12.30. Probably should update the version requirements, but I figured it would be safer to not start tweaking those, as the error message says.

mbercx replied:
However, this leads to version constraint conflicts:

```console
tf init
Initializing modules...
Downloading terraform-aws-modules/eks/aws 14.0.0 for eks...
- eks in .terraform/modules/eks
- eks.fargate in .terraform/modules/eks/modules/fargate
- eks.node_groups in .terraform/modules/eks/modules/node_groups
Downloading terraform-aws-modules/iam/aws 2.6.0 for iam_assumable_role_admin...
- iam_assumable_role_admin in .terraform/modules/iam_assumable_role_admin/modules/iam-assumable-role-with-oidc
Downloading terraform-aws-modules/vpc/aws 2.70.0 for vpc...
- vpc in .terraform/modules/vpc

Initializing the backend...

Initializing provider plugins...
- Checking for available provider plugins...
- Downloading plugin for provider "local" (hashicorp/local) 2.0.0...
- Downloading plugin for provider "null" (hashicorp/null) 3.0.0...

No provider "aws" plugins meet the constraint ">= 2.28.1,>= 3.10,>= 3.22.0,~> 2.23".

The version constraint is derived from the "version" argument within the
provider "aws" block in configuration. Child modules may also apply
provider version constraints. To view the provider versions requested by each
module in the current configuration, run "terraform providers".

To proceed, the version constraints for this provider must be relaxed by
either adjusting or removing the "version" argument in the provider blocks
throughout the configuration.

- Downloading plugin for provider "template" (hashicorp/template) 2.2.0...
- Downloading plugin for provider "kubernetes" (hashicorp/kubernetes) 1.11.4...
- Downloading plugin for provider "helm" (hashicorp/helm) 2.0.2...
- Downloading plugin for provider "random" (hashicorp/random) 3.0.1...

Error: no suitable version is available
```

This can be fixed by changing the version of `iam_assumable_role_admin` in `autoscaler.tf`. (`~>` -> `>=` i.e. updating the version requirements after all 😅). However, then the `apply` step fails:

```console
terraform apply -var-file your-cluster.tfvars 

Warning: Provider source not supported in Terraform v0.12

  on .terraform/modules/vpc/versions.tf line 5, in terraform:
   5:     aws = {
   6:       source  = "hashicorp/aws"
   7:       version = ">= 3.10"
   8:     }

A source was declared for provider aws. Terraform v0.12 does not support the
provider source attribute. It will be ignored.


Error: Invalid data source

  on efs.tf line 37, in data "helm_repository" "stable":
  37: data "helm_repository" "stable" {

The provider provider.helm does not support data source "helm_repository".
```

So yeah, switching back to v0.14.6 of terraform and re-initializing fixes the first error, but the second remains. This is because it was deprecated, see [the documentation of the helm provider v1.3.0](https://registry.terraform.io/providers/hashicorp/helm/1.3.0/docs/data-sources/repository)

mbercx replied:
As you may have noticed, the deprecation warning was not very helpful, but this was fixed [here](https://github.com/hashicorp/terraform-provider-helm/pull/491). Basically you no longer have to user the `helm_repository` data source and simply provide it as an attribute of the `helm_release`. When i do this (in both the `efs.tf` and `autoscaler.tf` files, there is only one error that remains:

```console
Error: Unsupported argument

  on main.tf line 29, in provider "kubernetes":
  29:   load_config_file       = false
```

Commenting out this line let's me run `terraform plan` successfully. But yeah, probably not a good idea to just swashbuckle my way through this and then deploy.

csadorf replied:
The specification of the required versions, including terraform itself have been updated.

mbercx commented on aws/terraform/README.md:
````diff
@@ -0,0 +1,286 @@
+You will also need this repo. You can get it with:
+
+```
+git clone https://github.com/pangeo-data/terraform-deploy.git
````
I don't understand why we are not using the terraform files in _this_ repo, but instead download those of the blog post. Is this copy-pasta? 🙃

csadorf replied:
Yup, I failed to update that. Should be fixed.

mbercx commented on aws/terraform/aws/autoscaler.tf:
```diff
@@ -0,0 +1,100 @@
+# Create IAM role + automatically make it available to cluster autoscaler service account
+module "iam_assumable_role_admin" {
+  source                        = "terraform-aws-modules/iam/aws//modules/iam-assumable-role-with-oidc"
+  version                       = "~> v2.6.0"
```
I think this should be changed to:

```suggestion
  version                       = ">= v2.6.0"
```

Else I get both issues with the terraform version (needs to downgrade), or version conflicts with the aws provider:

```console
No provider "aws" plugins meet the constraint ">= 2.28.1,>= 2.55.0,>= 3.10,~> 2.23".

The version constraint is derived from the "version" argument within the
provider "aws" block in configuration. Child modules may also apply
provider version constraints. To view the provider versions requested by each
module in the current configuration, run "terraform providers".

To proceed, the version constraints for this provider must be relaxed by
either adjusting or removing the "version" argument in the provider blocks
throughout the configuration.


Error: no suitable version is available
```

csadorf replied:
All versions have been updated (see other comment).

mbercx reviewed (commented):
I've been struggling to get this to work for a bit, and left some comments with some of the issues I've encountered. However, part of the issue was that I wasn't using the files in this repo, because I blindly followed the instructions for cloning the blog post repo like an idiot. 🤦 
(It was a Saturday morning, so give me some slack 😅)

There still seems to be some issue with the versions of the various providers. First it complained that my terraform version was not compatible, so i downgraded to 0.12.30. Next there were version conflicts for the `aws` provider. The `vpc` module requires `aws >= 3.10`, but the `iam_assumable_role_admin` module requires `aws ~> 2.23`. Downgrading the version of the `vpc` is not an option, this results in all sort of troubles. 

So, the simple solution, I think, is just to allow higher versions for the `iam_assumable_role_admin` module. This seems to fix all issues when using terraform v0.14.6. 

However, when trying to execute `terraform apply`, I get an error saying the Kubernetes cluster is unreachable. I tried it twice again, but to no avail. So then I just `terraform destroy` to terminate the instances that were already running (i.e. the eks cluster).

csadorf commented:
@yakutovicha More extensive changes will be needed, I'll reopen this PR once I consider it ready to merge.

csadorf committed:
Adapt README for AiiDAlab.

csadorf committed:
Reformat README file to follow 'one sentence per line' formatting style.

csadorf committed:
Update versions for terraform version 0.13.

csadorf committed:
Update aws/terraform/README.md .

csadorf committed:
Upgrade terraform to v0.14.6.

csadorf committed:
Update hashicorp/helm to version 2.0.2.

csadorf committed:
Reduce auto scaler group sizes.

csadorf commented:
We discovered a non-derterministic issue that sometimes leads to a failure of the destruction of all resources using the `terraform destroy` command in `aws/terraform/aws`.
The issues causes the following resources to remain:
```
$ terraform state list
helm_release.cluster-autoscaler
helm_release.efs-provisioner
kubernetes_namespace.support
module.eks.kubernetes_config_map.aws_auth[0]
```
It appears that these resources fail to be destroyed in cases where the kubernetes cluster is destroyed prior to these resources (race condition).

csadorf committed:
Update terraform/README.md

csadorf commented:
I think I would still like to merge this despite above mentioned issues with destruction. I am not sure whether I am going to be able to easily resolve it and the cluster setup is not blocked by this.

csadorf committed:
Restructure directory hierarchy.

 * Remove instructions for eksctl.

   It is not recommended to take that approach, but one should use
   terraform instead.

 * Move aws/terraform/ into terraform/ .

csadorf commented:
@mbercx Somehow GitHub does not allow me to re-request your review. But this is ready for your second review. ;)

mbercx commented on config-template.yaml:
```diff
@@ -1,11 +1,12 @@
 proxy:
     secretToken: "{{ secret_token }}"
+    https:
+      enabled: false
```
Eventually we set this to `true`, right? Could you explain again why we don't use HTTPS right away? I should probably also a note about this on the google doc.

csadorf replied:
Do you think that the documentation we are linking to [here](https://github.com/aiidalab/aiidalab-k8s/tree/add-aws-configuration#ssl-certificates-and-ingress) is insufficient?

mbercx replied:
No, right, I just had forgotten about this step. Looks good. 👍

mbercx commented on config-template.yaml:
```diff
@@ -1,11 +1,12 @@
     name: aiidalab/aiidalab-docker-stack
-    #tag: latest
-    tag: 20.02.2
+    tag: 20.12.0
```
Any reason why we're not using v21.02.1?

csadorf replied:
No, just wasn't released at the time. I'd say let's update that as part of a separate change set since it's not really relevant to this PR.

mbercx commented on terraform/README.md:
```diff
@@ -0,0 +1,214 @@
+The [documentation on configuring AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-configure.html) should help.
+
+The `aws-creds` configuration directory is used to setup a `terraform-bot` user that has exactly tailored permissions to setup the kubernetes infrastructure.
+It is not strictly necessary to use the bot user to create the cluster, however it is recommended to follow the principle of least privilege and instructions following are going to assume the use of the bot user.
```
```suggestion
It is not strictly necessary to use the bot user to create the cluster, however it is recommended to follow the principle of least privilege and the following instructions are going to assume the use of the bot user.
```

mbercx commented on terraform/README.md:
````diff
@@ -0,0 +1,214 @@
+terraform apply -var-file=aiidalab-cluster.tfvars
+```
+
+The initial setup of the infrastructure will take about 15 min or even more (the EKS cluster alon takes about 10 minutes to be created).
````
```suggestion
The initial setup of the infrastructure will take about 15 min or even more (the EKS cluster alone takes about 10 minutes to be created).
```

mbercx commented on terraform/aws-creds/.terraform.lock.hcl:
```diff
@@ -0,0 +1,22 @@
+# This file is maintained automatically by "terraform init".
```
Why do we leave this file in the repository? Just to be able to check if there is a `diff` when the file is generated by `terraform init`?

csadorf replied:
It is recommended to add this file to version control, see here: https://www.hashicorp.com/blog/terraform-0-14-introduces-a-dependency-lock-file-for-providers

mbercx replied:
Great, thanks!

mbercx reviewed (changes requested):
Looking good @csadorf! I just left a few minor comments and questions.

One more general question: Should we explain somewhere which resources to change to get more/less pods, depending on the number of participants? Not sure if this should be here or in the google doc.

csadorf commented:
@mbercx I responded to your comments.

mbercx reviewed (approved):
Thanks for replying to my questions, @csadorf!

csadorf commented:
@yakutovicha Let me know whether you would like to have a final look at this prior to merge.

csadorf committed:
Disable https by default.

csadorf committed:
Switch storage class to 'aws-efs' and 'gp2' for aws.

csadorf committed:
Revise custom config declaration.

Old style was deprecated.
